triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala

```diff
@@ -125,49 +125,70 @@ object Server extends StrictLogging {
+        requestId <- Try(UUID.fromString(state)).toOption
+        redirectUri <- requests.remove(requestId)
+      } yield redirectUri
+      redirectUri match {
+        case Some(redirectUri) => provide(redirectUri)
+        case None => complete(StatusCodes.NotFound)
+      }
+    }
+
+    concat(
+      parameters(('code, 'state ?))
+        .as[OAuthResponse.Authorize](OAuthResponse.Authorize) { authorize =>
+          popRequest(authorize.state) {
+            redirectUri =>
+              extractRequest {
+                request =>
+                  val body = OAuthRequest.Token(
+                    grantType = "authorization_code",
+                    code = authorize.code,
+                    redirectUri = toRedirectUri(request.uri),
+                    clientId = config.clientId,
+                    clientSecret = config.clientSecret)
+                  import com.daml.oauth.server.Request.Token.marshalRequestEntity
+                  val tokenRequest =
+                    for {
+                      entity <- Marshal(body).to[RequestEntity]
+                      req = HttpRequest(
+                        uri = config.oauthToken,
+                        entity = entity,
+                        method = HttpMethods.POST)
+                      resp <- Http().singleRequest(req)
+                      tokenResp <- if (resp.status != StatusCodes.OK) {
+                        Unmarshal(resp).to[String].flatMap { msg =>
+                          Future.failed(new RuntimeException(
+                            s"Failed to retrieve token at ${req.uri} (${resp.status}): $msg"))
+                        }
+                      } else {
+                        Unmarshal(resp).to[OAuthResponse.Token]
+                      }
+                    } yield tokenResp
+                  onSuccess(tokenRequest) { token =>
+                    setCookie(HttpCookie(cookieName, token.toCookieValue)) {
+                      redirect(redirectUri, StatusCodes.Found)
+                    }
                   }
-                } else {
-                  Unmarshal(resp).to[OAuthResponse.Token]
-                }
-              } yield tokenResp
-              onSuccess(tokenRequest) { token =>
-                setCookie(HttpCookie(cookieName, token.toCookieValue)) {
-                  redirect(redirectUri, StatusCodes.Found)
-                }
               }
           }
+        },
+      parameters(('error, 'error_description ?, 'error_uri.as[Uri] ?, 'state ?))
+        .as[OAuthResponse.Error](OAuthResponse.Error) { error =>
+          popRequest(error.state) { redirectUri =>
+            val uri = redirectUri.withQuery {
+              var params = redirectUri.query().to[Seq]
+              params ++= Seq("error" -> error.error)
+              error.errorDescription.foreach(x => params ++= Seq("error_description" -> x))
+              Uri.Query(params: _*)
+            }
+            redirect(uri, StatusCodes.Found)
```

Comment:
This is a change to the specification of the auth middleware. Before it didn't mention authorization failure in this context.

I've considered these alternatives to forwarding the error:
- Return an error from the auth middleware directly. This would not require any error handling code in the trigger service. However, it also doesn't allow any auth failure related cleanup or other dedicated handling in the auth service's client. E.g. in the trigger service we can delete the stored continuation.
- Just redirect but not pass an error parameter. This would also not require any error handling code in the trigger service. The stored continuation would still be popped. However, it wouldn't give the auth middleware's client the opportunity to react to login failure. Instead, it would only fail again at the next /auth request. The client can still choose to ignore the error parameter. Meaning, from the client's perspective this approach is still a possibility.

Reply:
Your approach seems very sensible. I would argue that it also matches the interaction between the oauth client and authorization server: If the user doesn’t pass the consent screen, you also get an error response on the client. That said, I think it’s probably not reasonable to rely on the error response always arriving, e.g., if he user just closes the tab so we probably want some timeout to clean things up in addition to this. But that’s definitely a separate issue.

Reply:
> That said, I think it’s probably not reasonable to rely on the error response always arriving, e.g., if he user just closes the tab so we probably want some timeout to clean things up in addition to this. But that’s definitely a separate issue.

Absolutely, that's in part what https://github.com/digital-asset/daml/issues/7721 is for. That and also DOS attacks.